larpandoracontent/LArUtility/RANSAC/RANSAC.h

```diff
@@ -0,0 +1,245 @@
+{
+
+template <class T, int t_numParams>
+class RANSAC
+{
+
+public:
+
+    /**
+     *  @brief  Default constructor.
+     *
+     *  @param threshold Threshold to be used for model to point distance.
+     *  @param numIterations How many iterations to run at most.
+     */
+    RANSAC(double threshold, int numIterations = 1000) :
+        m_numIterations(numIterations),
+        m_threshold(threshold)
+    {
+        this->Reset();
+    };
+
+    /**
+     *  @brief  Reset all the data.
+     */
+    void Reset() { m_data.clear(); };
+
+    /**
+     *  @brief Destructor
+     */
+    virtual ~RANSAC() {};
+
+    /**
+     *  @brief Get the best model, where the best model has the most inliers.
+     */
+    std::shared_ptr<T> GetBestModel() { return m_bestModel; };
+
+    /**
+     *  @brief Get the parameters that are inliers for the best model.
+     */
+    ParameterVector& GetBestInliers() { return m_bestInliers; };
+
+    /**
+     *  @brief Get the second best model, where the second best model has the second most unique inliers, compared to the best model.
+     *         This prevents the second best model being a model that is right next to the best model, and is instead somewhat more
+     *         unique.
+     */
+    std::shared_ptr<T> GetSecondBestModel() { return m_secondBestModel; };
+
+    /**
+     *  @brief Get the parameters that are inliers for the second best model.
+     */
+    ParameterVector& GetSecondBestInliers() { return m_secondBestInliers; };
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+    /**
+     *  @brief  Given a vector of data, get the best and second best model for it.
+     *
+     *  @param  data The data to use for this RANSAC run.
+     *
+     *  @return bool If the models were populated.
+     */
+    bool Estimate(const ParameterVector &data)
+    {
+        if (data.size() <= t_numParams)
+            return false;
+
+        m_data = data;
+        this->GenerateSamples();
+
+        std::vector<double> inlierFrac(m_numIterations, 0.0);
+        std::vector<ParameterVector> inliers(m_numIterations);
+        std::vector<std::shared_ptr<T>> sampledModels(m_numIterations);
+
+        RANSAC::CheckModel(inlierFrac, inliers, sampledModels);
+        double bestModelScore(-1);
+
+        for (unsigned int i = 0; i < sampledModels.size(); ++i)
+        {
+            if (inlierFrac[i] == 0.0)
+                continue;
+
+            if (inlierFrac[i] > bestModelScore)
+            {
+                bestModelScore = inlierFrac[i];
+                m_bestModel = sampledModels[i];
+                m_bestInliers = inliers[i];
+            }
+        }
+
+        double secondModelScore(-1);
+
+        for (unsigned int i = 0; i < sampledModels.size(); ++i)
+        {
+            if (inlierFrac[i] == 0.0)
+                continue;
+
+            if (inlierFrac[i] > (secondModelScore * 0.9))
```

Comment:
Would it make sense to put 0.9 into a variable, or do you anticipate this will never be changed?

Reply:
(Copying a comment over from Slack for context)

As for making it a variable, I wasn’t too sure. I think the set I ended up on was ones that would be somewhat easily tuned by someone, whereas I think that 0.9 requires a lot of internal knowledge about how the method works, so in reality wouldn’t be easily tuned.

(That is compared to something like “how far in front of fit do you look”, this is a lot more abstract of a variable to think about)